apps/web/src/components/BlocksEditor/Editor/state/promptlToLexical/fromAstToBlocks.ts

```diff
@@ -563,6 +602,163 @@ function trimEmptyBlocks(
+  code.children = code.children.map((child) => {
+    const normalized = child.text
+      .split('\n')
+      .map((line) => (line.startsWith('  ') ? line.slice(2) : line))
+      .join('\n')
+    return { ...child, text: normalized }
+  })
+}
+
+function normalizeIndentationInMessageChildren(
+  blocks: (ParagraphBlock | CodeBlock)[],
+): (ParagraphBlock | CodeBlock)[] {
+  blocks.forEach((block) => {
+    if (block.type === 'paragraph') normalizeParagraphIndentation(block)
+    if (block.type === 'code') normalizeCodeBlockIndentation(block)
+  })
+  return blocks
+}
+
+function normalizeIndentationInStepChildren(blocks: StepChild[]): StepChild[] {
+  blocks.forEach((block) => {
+    if (block.type === 'paragraph') normalizeParagraphIndentation(block)
+    if (block.type === 'code') normalizeCodeBlockIndentation(block)
+  })
+  return blocks
+}
+
+function trimLeadingMessageChildren(
+  blocks: (ParagraphBlock | CodeBlock)[],
+): (ParagraphBlock | CodeBlock)[] {
+  for (let i = 0; i < blocks.length; i++) {
+    const block = blocks[i]!
+
+    if (block.type != 'paragraph') {
+      return blocks.slice(i)
+    }
+
+    const trimmed = trimLeadingTexts(block.children)
+    if (trimmed.length > 0) {
+      block.children = trimmed
+      return blocks.slice(i)
+    }
+  }
+
+  return []
+}
+
+function trimTrailingMessageChildren(
+  blocks: (ParagraphBlock | CodeBlock)[],
+): (ParagraphBlock | CodeBlock)[] {
+  for (let i = blocks.length - 1; i >= 0; i--) {
+    const block = blocks[i]!
+
+    if (block.type != 'paragraph') {
+      return blocks.slice(0, i + 1)
+    }
+
+    const trimmed = trimTrailingTexts(block.children)
+    if (trimmed.length > 0) {
+      block.children = trimmed
+      return blocks.slice(0, i + 1)
+    }
+  }
+
+  return []
+}
+
+function trimEmptyMessageChildren(
+  blocks: (ParagraphBlock | CodeBlock)[],
+): (ParagraphBlock | CodeBlock)[] {
+  return trimTrailingMessageChildren(trimLeadingMessageChildren(blocks))
+}
+
+function trimLeadingStepChildren(blocks: StepChild[]): StepChild[] {
+  for (let i = 0; i < blocks.length; i++) {
+    const block = blocks[i]!
+
+    if (block.type != 'paragraph') {
+      return blocks.slice(i)
+    }
+
+    const trimmed = trimLeadingTexts(block.children)
+    if (trimmed.length > 0) {
+      block.children = trimmed
+      return blocks.slice(i)
+    }
+  }
+
+  return []
+}
+
+function trimTrailingStepChildren(blocks: StepChild[]): StepChild[] {
+  for (let i = blocks.length - 1; i >= 0; i--) {
+    const block = blocks[i]!
+
+    if (block.type != 'paragraph') {
+      return blocks.slice(0, i + 1)
+    }
+
+    const trimmed = trimTrailingTexts(block.children)
+    if (trimmed.length > 0) {
+      block.children = trimmed
+      return blocks.slice(0, i + 1)
+    }
+  }
+
+  return []
+}
+
+function trimEmptyStepChildren(blocks: StepChild[]): StepChild[] {
+  return trimTrailingStepChildren(trimLeadingStepChildren(blocks))
+}
```

Comment:
Move all of this to utils. This file should be as easy to understand as possible.